sycl/include/CL/sycl/handler.hpp

```diff
@@ -1048,37 +1053,26 @@ class __SYCL_EXPORT handler {
     shared_ptr_class<detail::queue_impl> QueueCopy = MQueue;
     this->finalize();
 
-    // 2. Run the additional aux kernel as many times as needed to reduce
-    // all partial sums into one scalar.
+    // 2. Find the maximal work group size usable for the additional kernel(s).
+    size_t MaxWGSize;
+    if (NWorkGroups > 1) {
+      constexpr bool HFR = Reduction::has_fast_reduce;
+      size_t OneElemSize = HFR ? 0 : sizeof(typename Reduction::result_type);
+      MaxWGSize = intel::detail::reduGetMaxWGSize(QueueCopy, OneElemSize);
+      assert(MaxWGSize > 1 &&
```

Comment:
I think this is fine for now, because the new behavior is an improvement and it's pretty unlikely that this assertion will ever trigger.  But we might want to consider introducing a fallback path that always works with 1 work-item, just by iterating over all the partial results sequentially.  This would be terribly slow and we'd probably want to issue a performance warning, but at least there would be no cases that didn't work.

Reply:
Thank you. Your comment made me think that the main kernel also needs that safety check.
The only bad think in such check is that it does nothing in all cases (except pretty crazy test cases), but does take time to it. Perhaps, it is really minor overhead and that the info requests to device work really fast.

Reply:
I'm not sure I follow what you mean about the main kernel.  The main kernel is the one that the user provided, plus the reduction prologue/epilogue code, right?

Are you suggesting that if the user provides an ND-range of 1 work-item, you could just perform a serial reduction?  Or something else?

Reply:
For this scenario:
a) User wants nd_range<1>{16,16}
b) Reduction::result_type is let's say 32Kb, and thus MaxWGSize == 2 (because the local memory is 64Kb)
the local accessor at the line https://github.com/intel/llvm/blob/sycl/sycl/include/CL/sycl/intel/reduction.hpp#L652 cannot be used as it would require 16*16Kb

Handling nd_range<1>{16,16} would require a kernel that runs user's lambda, but does not do the reduction/loop https://github.com/intel/llvm/blob/sycl/sycl/include/CL/sycl/intel/reduction.hpp#L678
instead it would simply write Reducer::MValue to global accessor/array with partial sums.